### ISSUE TITLE
[Question] What does `fio: cache invalidation of` mean?

### ISSUE THREAD

**Issue by boknowswiki** [axboe/fio#1060]
Hi,

When I run `fio` in a k8s environment, I seen this error message:
`fio: cache invalidation of /dev/XXX failed: Permission denied`
And I have run this container as user root.

My question is what's the impact of this error message? The tests seems keep going and I got some test results. Just would like to know the impact. Please help me to understand.

Thanks,
Bo

**Comment by sitsofe** [comment]
TLDR; The impact is your results may be unreliable due to distortion of what's already exists in the Linux page cache. If you're trying to benchmark a raw block device it is is wiser to so outside of a container.

Longer answer: Linux has a [page cache](https://en.wikipedia.org/wiki/Page_cache) and fio couldn't tell the kernel to drop its contents before fio started its run. This is important to know because the state of the cache can massively distort benchmarking. For example, if what you are reading is already in the cache and is serviced by it you can get a speed far greater than if you actually sent to the disk and this is just one scenario (there are more). One extreme situation can be that you just end up testing the speed of your cache...

It's possible your container is blocking various syscalls - just because you're root doesn't mean the container doesn't have seccomp policy attached (https://docs.docker.com/engine/security/seccomp/ ). However identifying/proving this and solving  the issue is beyond the remit of the fio project...

@boknowswiki Does that explain things from fio's standpoint?

**Comment by boknowswiki** [comment]
@sitsofe, yes, it's very clear to me. Thanks a lot. Let me close this question.

**Comment by boknowswiki** [comment]
@sitsofe , sorry, one more question from FIO side:
If I set direct=1, which `If true, use non-buffered I/O (usually O_DIRECT). Default: false.`, my understand in this case, we can ignore the cache invalidation issue, since I don't use cache any way. Is this correct? And in this case the test should be valid and consistent over times, right? Thanks.

**Comment by axboe** [comment]
direct=1 will still attempt to invalidate the page cache, since O_DIRECT is more efficient if we don't have any page cache entries in that range. But that'll just show itself as using a bit more CPU and perhaps running a bit slower if you've maxed out the CPU doing IO, it won't have any other ill effects.

**Comment by boknowswiki** [comment]
@axboe , since direct=1 will invalidate the cache, and due to the permission denied, does that mean my test is still using the cache if there is any cache hit? Thanks.

**Comment by axboe** [comment]
It won't use the cache if direct=1 is set, but it will slow down the O_DIRECT path because the pages exist in the cache due to the failed invalidation.

**Comment by boknowswiki** [comment]
@axboe , I see. Thanks for your explanation.
